### PR TITLE
Make entity ordering in state representations consistent

### LINKED ISSUE #8623: Random Core issues when training the model, when using AugmentedMemoizationPolicy
**Rasa version**:
2.4.5
**Rasa SDK version** (if used & relevant):

**Python version**:
3.7.9
**Operating system** (windows, osx, ...):
mac 10.15.7
**Issue**:

Normally when training this story AugmentedMemoizationPolicy should always kick in, we have noticed that after retraining and trying to do this story, it fails to use augmentation policy on the 3rd step once in a while. We checked the hashes and noticed that every few builds of the model these change for some reason. 

**Command or request that led to error**:
```
rasa train --augmentation 0

```

**Content of configuration file (config.yml)** (if relevant):
```yml
# Configuration for Rasa Core.
policies:
  - name: AugmentedMemoizationPolicy
    max_history: 6
  - name: TEDPolicy
    max_history: 6
    epochs: 30
  - name: FormPolicy
  - name: RulePolicy
    core_fallback_threshold: 0.3
    core_fallback_action_name: "action_fallback_core"
    enable_fallback_prediction: True
    check_for_contradictions: True

**Content of domain file (domain.yml)** (if relevant):
```yml
slot_one is a categorical one

```
training story
```
version: "2.0"
stories:
- story: example story
  steps:
  - intent: intent_one
  - action: action_one
  - slot_was_set:
    - slot_one: 1+
  - action: utter_one
  - intent: intent_two
    entities:
    - entity_one: entity_one_value
    - entity_two: entity_two_value
  - action: action_three

```

### PR DESCRIPTION
**Pull request by samsucik**
**Proposed changes**:
- Explicitly order entities in state representations. Previously, user utterances with multiple entities could lead to memoisation policies defaulting: if the order of the entities was different from that saved in the policy's memorised state-action lookup, the policy wouldn't find an applicable memorised "rule" to follow.
- Fixes #8623 

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] ~updated the documentation~
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by samsucik** [comment]
I've added a simple bot for manual testing. When I set `PYTHONHASHSEED=1` and train the bot (`rasa train`), then change to `PYTHONHASHSEED=4` and run `rasa test`, the memoization policy fails to correctly predict the test stories. Without changing the seed, the policy succeeds.

**Comment by samsucik** [comment]
Right now, all tests are passing but the newly added one always fails -- as expected, since the fix isn't enabled yet. I'm gonna enable it now.

File: tests/shared/core/test_domain.py [hunk 1]
```diff
@@ -1110,6 +1112,127 @@ def test_get_featurized_entities():
...
+        "x",
+        "h",
+        "v",
+        "s",
+        "o",
+        "a",
+        "t",
+        "g",
+        "r",
+    ]
```
[the remark this answers is not in the document]
Comment (samsucik) [reply tests/shared/core/test_domain.py:1211]: I agree, I just didn't want to bring even more randomness into the test case 😆  But I guess I'll change it now.

[the remark this answers is not in the document]
Comment (twerkmeister) [reply tests/shared/core/test_domain.py:1211]: I think it's not even necessary in the end, if you don't shuffle things yourself, with the old behavior they would come out shuffled. I guess you did it to just show the behavior even more explicitly, right?

**Pull request merged**